hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyRenameRequestWithFSO.java

```diff
@@ -243,27 +244,56 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
   }
 
   @SuppressWarnings("parameternumber")
-  private OMClientResponse renameKey(long toKeyParentId,
-      long trxnLogIndex, OmKeyInfo fromKeyValue, boolean isRenameDirectory,
-      String toKeyName, long modificationTime, OMResponse.Builder omResponse,
-      OzoneManager ozoneManager) throws IOException {
-
+  private OMClientResponse renameKey(OmKeyInfo toKeyParent, String toKeyName,
+      OmKeyInfo fromKeyValue, String fromKeyName, boolean isRenameDirectory,
+      long modificationTime, OzoneManager ozoneManager,
+      OMResponse.Builder omResponse, long trxnLogIndex) throws IOException {
     final OMMetadataManager ommm = ozoneManager.getMetadataManager();
     final long volumeId = ommm.getVolumeId(fromKeyValue.getVolumeName());
     final long bucketId = ommm.getBucketId(fromKeyValue.getVolumeName(),
             fromKeyValue.getBucketName());
     final String dbFromKey = ommm.getOzonePathKey(volumeId, bucketId,
             fromKeyValue.getParentObjectID(), fromKeyValue.getFileName());
     String toKeyFileName = OzoneFSUtils.getFileName(toKeyName);
+    OmKeyInfo fromKeyParent = null;
+    OMMetadataManager metadataMgr = ozoneManager.getMetadataManager();
+    Table<String, OmDirectoryInfo> dirTable = metadataMgr.getDirectoryTable();
 
     fromKeyValue.setUpdateID(trxnLogIndex, ozoneManager.isRatisEnabled());
     // Set toFileName
     fromKeyValue.setKeyName(toKeyFileName);
     fromKeyValue.setFileName(toKeyFileName);
     // Set toKeyObjectId
-    fromKeyValue.setParentObjectID(toKeyParentId);
-    //Set modification time
-    fromKeyValue.setModificationTime(modificationTime);
+    if (toKeyParent != null) {
+      fromKeyValue.setParentObjectID(toKeyParent.getObjectID());
+    } else {
+      String bucketKey = metadataMgr.getBucketKey(
+          fromKeyValue.getVolumeName(), fromKeyValue.getBucketName());
+      OmBucketInfo omBucketInfo =
+          metadataMgr.getBucketTable().get(bucketKey);
+      fromKeyValue.setParentObjectID(omBucketInfo.getObjectID());
```

Comment:
Do we want a rename immediately under the bucket to update the bucket's modification time? If yes, then `KeyRenameRequest` should be modified for consistent behavior on object store buckets, the corresponding response updated to flush the bucket to the DB, and tests added.

I'm not sure which option is better, but we should make a definitive call here. For context it looks like we are hardly ever setting bucket modification time except for changing the owner.

Reply:
It depends on how we view the volume and bucket, based on the ACL process of checking permissions, the volume and bucket can be treated as the upper-level directory of the key.

We currently will modify the bucket's modification time in set a bucket property, which lets us know when the bucket's properties were last modified, which is easy to understand. Would it be strange if the bucket's modification time was modified when the key was renamed?

Reply:
In FSO world, a bucker is root of the filesystem and should follow the same semantics. 
```
[root@nightly-71x-yj-1 ~]# cd /
[root@nightly-71x-yj-1 /]# stat /
  File: ‘/’
  Size: 4096      	Blocks: 8          IO Block: 4096   directory
Device: fd10h/64784d	Inode: 1835009     Links: 25
Access: (0755/drwxr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2021-05-02 08:23:05.977173911 +0000
Modify: 2022-11-28 06:27:06.659388314 +0000
Change: 2022-11-28 06:27:06.659388314 +0000
 Birth: -
[root@nightly-71x-yj-1 /]# touch file
[root@nightly-71x-yj-1 /]# stat /
  File: ‘/’
  Size: 4096      	Blocks: 8          IO Block: 4096   directory
Device: fd10h/64784d	Inode: 1835009     Links: 25
Access: (0755/drwxr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2021-05-02 08:23:05.977173911 +0000
Modify: 2022-11-28 19:02:40.442335928 +0000
Change: 2022-11-28 19:02:40.442335928 +0000
 Birth: -
[root@nightly-71x-yj-1 /]# mv file file2
[root@nightly-71x-yj-1 /]# stat /
  File: ‘/’
  Size: 4096      	Blocks: 8          IO Block: 4096   directory
Device: fd10h/64784d	Inode: 1835009     Links: 25
Access: (0755/drwxr-xr-x)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2021-05-02 08:23:05.977173911 +0000
Modify: 2022-11-28 19:02:49.217171265 +0000
Change: 2022-11-28 19:02:49.217171265 +0000
 Birth: -
```

Reply:
Yes, volume modification time should change on bucket creation, in my opinion.

Reply:
This is an interesting discussion 😄 
It seems like we are leaning towards the following:
- create bucket changes volume modification time regardless of bucket type.
- create/rename/delete file or directory at the root of a bucket changes bucket modification time for FSO buckets.

I wonder if we should also have object create/rename/delete in OBS buckets change the bucket modification time? I am kind of thinking no. Since OBS buckets do not have filesystem semantics the bucket is a separate entity. However I do not have a strong opinion as long as we have some justification for the approach we decide on.

Reply:
I have a new commit that implements this feature.